README.md
# Escape the Burning Space Station

Escape the Burning Space Station is a Python text-based adventure game.
In it you will fight unique monsters, encounter obstacles and escape the burning room
Target audience: Teenage audiences

## Installation

Python 3.8 is needed to play this game. Go to the [Python downloads page](https://www.python.org/downloads/) to install Python.

## Run the Program

```bash
python main.py
```

## How to play

To play the game read the text and enter the number of the action for your player to do


## How to quit

you can quit by either
- Inputting 4 to the menu
- Closing the program
- Pressing CTRL + C

## Tips

- Don't be scared to fight enemies
- Equip your shield to cut incoming damage in half
- Use healing wisely

CHANGELOG.md
# RPG: Escape the Burning Space Station Changelog
## v1.0
- Added movement
- Added playable map(Space Station)
- Added class: Player

## v1.1
- Players can no longer phase through fire
- Enemies now appear on the map
- Escape pod key is now obtainable

## v2
- Added combat
- Wall of fire is now passable by defeating all enemies and accessing the fire extinguisher

## v2.1
- Fixed bug that would skip player turn in combat if invalid input was entered

## v2.2
- Fixed bug where shield would not unequip after combat ended

## v2.1.1
- Fixed bug where number of bandages remaining did not display after healing
- Fixed bug where players could win without having grabbed the key

char.py
# Course: CS 30
# Period: 1
# Date created: 21/04/10
# Date last modified: 21/04/15
# Name: Ahmed Keshta
# Description: Player data module for rpg game


# starting position for player
startPos = [0, 0]


# defining starting variables for player
maxHealth = 100
name = input(f'What is your name:\n')
x = startPos[0]
y = startPos[1]
health = 100
shield = 0
combat = 0


# creating class for player
class Player:
    def __init__(self, name, health, maxHealth, x, y, shield):
        self.health = health
        self.maxHealth = maxHealth
        self.shield = shield
        self.name = name
        self.x = x
        self.y = y

    def heal(self, amount):
        '''
        heal function for player class using bandages
        '''
        self.health += amount


player = Player(name, health, maxHealth, x, y, shield)


data.py
# Course: CS 30
# Period: 1
# Date created: 21/03/11
# Date last modified: 21/03/16
# Name: Ahmed Keshta
# Description: Data module for game


# stats for entities
player = {'name': 'bill', 'health': 100, 'strength': 10, 'hunger': 9}
alien = {'name': 'alien', 'health': 100, 'damage': 10, 'points': 15}
spider = {'name': 'spider', 'health': 10, 'damage': 4, 'points': 5}
zombie = {'name': 'zombie', 'health': 50, 'damage': 8, 'points': 10}


# Item names, descriptions and stats
sword = {
    'name': 'Sword',
    'damage': 10,
    'description': 'Useful for close range enemies'
}


shield = {
    'name': 'Shield',
    'description': 'Reduces incoming damage when equipped'
}


bandage = {
    'name': 'Bandage',
    'description': 'Restores 10 hp to player',
    'heal': 10,
    'amount': 10
}


inventory = [sword, shield, bandage]
enemies = [alien, spider, zombie]
inv = []


for item in inventory:
    inv.append(item['name'])


class Inventory:
    def __init__(self, items):
        self.items = inventory


main.py
# Course: CS 30
# Period: 1
# Date created: 21/03/11
# Date last modified: 21/03/16
# Name: Ahmed Keshta
# Description: RPG Game Mock


import maps
import char
import UI
import classes
from classes import key
from UI import menu
from UI import inCombat


# variables to make life easier
clear = '\n' * 200
win = 0
name = char.player.name.title()


print(f'\nHello {name}, you are stuck in a burning space')
print(f'station and must make your way to the escape pod(◨).')
print(f'Watch out for tiles on fire, they will look like')
print(f'this "⧮". Enemies appear as E. You are here ⯐.')
input(f'Input anything to enter:\n')
print(clear)


# loop for running through program until win condition is achieved
while win < 1:
    maps.spaceStation()
    print(f'\n{name} has {char.player.health} hit points\n')
    if char.combat > 0:
        inCombat()
    
    elif char.combat == 0:
        menu()

    if char.player.x == classes.fire.x and classes.fire.check == 1:
        print('You can\'t go there without a fire extinguisher!'
        '\nObtain one by killing all 3 enemies')
        char.player.x = char.player.x - 1

    if char.player.x == classes.spider.x and char.player.y == classes.spider.y:
        char.combat = 1
        print('A giant spider jumped at you from the dark!')
        classes.spider.y = 20
        print(char.combat)

    if char.player.x == classes.alien.x and char.player.y == classes.alien.y:
        char.combat = 2
        print('A bloodthirsty alien ambushed you!')
        pass

    if char.player.x == classes.zombie.x and char.player.y == classes.zombie.y:
        char.combat = 3
        print('A hungry zombie is right in front of you!')
        classes.zombie.y = 20
        pass

    if char.player.x == key.x and char.player.y == key.y:
        key.grab()

    if char.player.x == 9 and char.player.y == 4 and key.check == 1:
        win = 3
    
    if char.player.health <= 0:
        print('You lose!')
        exit()

    if UI.death == 3:
        UI.death += 1
        print('You obtained the fire extinguisher and put out the fire!')
        classes.fire.ext()


print(f'Congratulations! You have escaped the burning space station!')


maps.py
# Course: CS 30
# Period: 1
# Date created: 21/04/10
# Date last modified: 21/04/15
# Name: Ahmed Keshta
# Description: Map module for RPG game


import char
import classes
from classes import key


# defining function for space station map
def spaceStation():
    '''
    Function for printing map and tiles
    '''
    separator = " "
    print(f'\n\n\n')
    a = ['▢', '▢', '▢', '▢', '▢', '▢', '▢', '▢', '▢', '▢']
    b = ['▢', '▢', '▢', '▢', '▢', '▢', '▢', '▢', '▢', '▢']
    c = ['▢', '▢', '▢', '▢', '▢', '▢', '▢', '▢', '▢', '▢']
    d = ['▢', '▢', '▢', '▢', '▢', '▢', '▢', '▢', '▢', '▢']
    e = ['▢', '▢', '▢', '▢', '▢', '▢', '▢', '▢', '▢', '◨']
    grid = [a, b, c, d, e]
    if classes.spider.y < 5:
        grid[classes.spider.y][classes.spider.x] = classes.spider.marker
    if classes.alien.y < 5:
        grid[classes.alien.y][classes.alien.x] = classes.alien.marker
    if classes.zombie.y < 5:
        grid[classes.zombie.y][classes.zombie.x] = classes.zombie.marker
    grid[key.y][key.x] = key.marker
    if char.player.y > 4:
        char.player.y = 4
    grid[char.player.y][char.player.x] = '⯐'
    for row in grid:
        grid[0][classes.fire.x] = classes.fire.marker
        grid[1][classes.fire.x] = classes.fire.marker
        grid[2][classes.fire.x] = classes.fire.marker
        grid[3][classes.fire.x] = classes.fire.marker
        grid[4][classes.fire.x] = classes.fire.marker
        print(separator.join(row))


menu.py
# Course: CS 30
# Period: 1
# Date created: 21/04/10
# Date last modified: 21/04/15
# Name: Ahmed Keshta
# Description: Menu module for RPG game


import player
import classes
import data


h = player.hunger
mh = player.maxHung
y = player.y
x = player.x
clear = '\n' * 200


# defining function for printing inventory
def printInv():
    '''
    Function for printing inventory
    '''
    separator = " "
    print(clear)
    print('Inventory:')
    print(f'\n{separator.join(data.inv)}\n')


# creating menu function for selecting actions
def menu():
    '''
    Menu for space station game
    '''
    print('Available options:')
    print('1) Move')
    print('2) Inventory')
    print('3) Use item')
    print('4) Quit')
    choice = input(f'What would you like to do?\n')


# creating option to move around the map
    if choice == "1" or choice.lower() == 'move':
        print('Where would you like to move:')
        print('1) Up')
        print('2) Down')
        print('3) Left')
        print('4) Right')
        choice = input()

        if choice == '1' and y > 0 or choice.lower() == 'up' and y > 0:
            print(clear)
            player.y = player.y - 1
            choice = ''

        elif choice == '2' and y < 4 or choice.lower() == 'down' and y < 4:
            print(clear)
            player.y = player.y + 1
            choice = ''

        elif choice == '3' and x > 0 or choice.lower() == 'left' and x > 0:
            print(clear)
            player.x = player.x - 1
            choice = ''

        elif choice == '4' and x < 9 or choice.lower() == 'right' and x < 9:
            print(clear)
            player.x = player.x + 1
            choice = ''

        else:
            print(clear)
            print("That is not a valid option!")

# creating option to view inventory
    elif choice == "2" or choice.lower() == 'inventory':
        printInv()

# creating option to use an item
    elif choice == "3" or choice.lower() == 'use an item':
        a = 1
        for item in data.inv:
            print(f'{a}) {item.title()}')
            a += 1
        choice = input(f'What would you like to use?\n')
        if choice == '1' or choice.lower() == 'sword':
            if player.combat == 1:
                print(clear)
                print(f'{player.name.title()} used their Sword')
                classes.Enemy.health = classes.Enemy.health - data.sword('damage')
            else:
                print(clear)
                print('You cannot use this while not in combat')
            choice = ''

        elif choice == '2' or choice.lower() == 'shield':
            if player.combat == 1:
                print(clear)
                print(f'{player.name.title()} used their Shield')
                player.shield = 1
            else:
                print(clear)
                print('You cannot use this while not in combat')
            choice = ''

        elif choice == '3' or choice.lower() == 'bandage':
            print(clear)
            if player.health < player.maxHealth:
                player.heal(data.bandage('heal'))
                if player.health > player.maxHealth:
                    player.health = player.maxHealth
                
            else:
                print(f'{player.name.title()} is already at maximum health')
            choice = ''

        elif choice == '4' or choice.lower() == 'apple':
            print(clear)
            if h < mh:
                player.hunger = player.hunger + data.apple('hunger')
                if h > mh:
                    h = mh
            else:
                print(f'{player.name.title()} is full')
            choice = ''

        else:
            print(clear)
            print("That is not a valid option!")

# creating option to quit game
    elif choice == "4" or choice.lower() == 'quit':
        print(clear)
        exit()

    else:
        print(clear)
        print("That is not a valid option!")


player.py
# Course: CS 30
# Period: 1
# Date created: 21/04/10
# Date last modified: 21/04/15
# Name: Ahmed Keshta
# Description: Player data module for rpg game


import data


# starting position for player
startPos = [0, 0]


# defining starting variables for player
maxHealth = 100
name = input(f'What is your name:\n')
x = startPos[0]
y = startPos[1]
health = 100
shield = 0
hunger = 10
maxHung = 10
combat = 0


# creating class for player
class Player:
    def __init__(self, name, health, maxHealth, x, y, shield, hunger, maxHung):
        self.health = 100
        self.name = name
        self.x = x
        self.y = y

    def heal(self, amount):
        '''
        heal function for player class using bandages
        '''
        self.health += amount


player = Player(name, health, maxHealth, x, y, shield, hunger, maxHung)


classes.py
# Course: CS 30
# Period: 1
# Date created: 21/04/22
# Date last modified: 21/04/26
# Name: Ahmed Keshta
# Description: Class module for RPG game


import data
from data import inv


# Enemy marker
enMarker = 'E'


# Enemy death count(when equal to 3 player obtains fire extinguisher)
death = 0


# Enemy base class
class Enemy:
    def __init__(self, name, health, x, y, damage, marker):
        self.health = 100
        self.name = name
        self.x = x
        self.y = y
        self.damage = damage
        self.marker = 'E'


# Map tile base class
class mapTile:
    def __init__(self, marker):
        self.marker = marker


# Class for default map tiles
class defTile(mapTile):
    def __init__(self, marker):
        self.marker = '▢'


# tile class containing key
class keyTile(mapTile):

    def __init__(self, marker, x, y, check):
        self.marker = marker
        self.x = x
        self.y = y
        self.check = check

    def grab(self):
        '''
        Function for when key is grabbed
        '''
        self.check = 1
        self.marker = '▢'
        print('You grabbed the key to the escape pod!')
        inv.append('key')


# tile class for tiles on fire
class Fire(mapTile):
    def __init__(self, check, x):
        self.x = x
        self.check = check
        self.marker = '⧮'

    def ext(self):
        self.check = 0
        self.marker = '▢'


# spider enemy class
class Spider(Enemy):
    def __init__(self, x, y, health, marker):
        self.health = health
        self.x = x
        self.y = y
        self.marker = marker

    def kill(self):
        self.marker = '▢'
        self.y = 20


# alien enemy class
class Alien(Enemy):
    def __init__(self, x, y, health, marker):
        self.health = health
        self.x = x
        self.y = y
        self.marker = marker

    def kill(self):
        self.marker = '▢'
        self.y = 20


# zombie enemy class
class Zombie(Enemy):
    def __init__(self, x, y, health, marker):
        self.health = health
        self.x = x
        self.y = y
        self.marker = marker

    def kill(self):
        self.marker = '▢'
        self.y = 20


# creation of enemy objects
spider = Spider(2, 1, data.spider['health'], enMarker)
alien = Alien(3, 3, data.alien['health'], enMarker)
zombie = Zombie(1, 3, data.zombie['health'], enMarker)


# creation of key tile object
key = keyTile('⚿', 3, 4, 0)


# creation of fire tile object
fire = Fire(1, 5)


UI.py
# Course: CS 30
# Period: 1
# Date created: 21/04/10
# Date last modified: 21/04/15
# Name: Ahmed Keshta
# Description: Menu module for RPG game

import char
import classes
import data


y = char.player.y
x = char.player.x
clear = '\n' * 200
death = 0
name = char.player.name.title()
bandage = data.bandage['amount']


# defining function for printing inventory
def printInv():
    '''
    Function for printing inventory
    '''
    separator = " "
    print(clear)
    print('Inventory:')
    print(f'\n{separator.join(data.inv)}\n')


# creating menu function for selecting actions
def menu():
    '''
    Menu for space station game
    '''
    global name
    global y
    global x
    global bandage

    print('Available options:')
    print('1) Move')
    print('2) Inventory')
    print('3) Use item')
    print('4) Quit')
    choice = input(f'What would you like to do?\n')

    # creating option to move around the map
    if choice == "1" or choice.lower() == 'move':
        print('Where would you like to move:')
        print('1) Up')
        print('2) Down')
        print('3) Left')
        print('4) Right')
        choice = input('\nEnter the number of your choice:')

        if choice == '1' and y > 0 or choice.lower() == 'up' and y > 0:
            print(clear)
            char.player.y = char.player.y - 1
            y = char.player.y
            choice = ''

        elif choice == '2' and y < 4 or choice.lower() == 'down' and y < 4:
            print(clear)
            char.player.y = char.player.y + 1
            y = char.player.y
            choice = ''

        elif choice == '3' and x > 0 or choice.lower() == 'left' and x > 0:
            print(clear)
            char.player.x = char.player.x - 1
            x = char.player.x
            choice = ''

        elif choice == '4' and x < 9 or choice.lower() == 'right' and x < 9:
            print(clear)
            char.player.x = char.player.x + 1
            x = char.player.x
            choice = ''

        else:
            print(clear)
            print("That is not a valid option!")

# creating option to view inventory
    elif choice == "2" or choice.lower() == 'inventory':
        printInv()

# creating option to use an item
    elif choice == "3" or choice.lower() == 'use an item':
        a = 1
        for item in data.inv:
            print(f'{a}) {item.title()}')
            a += 1
        choice = input(f'What would you like to use?\n')
        if choice == '1' or choice.lower() == 'sword':
            if char.combat == 1:
                print(clear)
                print(f'{name} used their Sword')
                classes.Enemy.health = classes.Enemy.health - data.sword(
                    'damage')
            else:
                print(clear)
                print('You cannot use this while not in combat')
            choice = ''

        elif choice == '2' or choice.lower() == 'shield':
            if char.combat == 1:
                print(clear)
                print(f'{name} used their Shield')
                char.player.shield = 1
            else:
                print(clear)
                print('You cannot use this while not in combat')
            choice = ''

        elif choice == '3' or choice.lower() == 'bandage':
            print(clear)
            if data.bandage['amount'] > 0:
                if char.player.health < char.player.maxHealth:
                    char.player.heal(data.bandage['heal'])
                    data.bandage['amount'] -= 1
                    bandage = data.bandage['amount']
                    if bandage == 0:
                        data.inv.remove('bandage')
                    print(f'You have {bandage} bandages remaining!')
                    if char.player.health > char.player.maxHealth:
                        char.player.health = char.player.maxHealth

                else:
                    print(
                        f'{name} is already at maximum health')

            else:
                print('You do not have any more bandages!')
            choice = ''

        else:
            print(clear)
            print("That is not a valid option!")


# creating option to quit game
    elif choice == "4" or choice.lower() == 'quit':
        print(clear)
        exit()

    else:
        print(clear)
        print("That is not a valid option!")


# Menu during a combat engagement
def fight():
    print('What would you like to do?')
    print('1) Swing sword')
    print('2) Equip/Unequip shield')
    print('3) Use bandages')
    print('4) Quit')
    choice = input('\nEnter the number of your choice:')
    print(clear)
    if choice == '1':
        print('You swung your sword!')
        if char.combat == 1:
            classes.spider.health -= 10

        if char.combat == 2:
            classes.alien.health -= 10

        if char.combat == 3:
            classes.zombie.health -= 10

    elif choice == '2':
        if char.shield == 0:
            char.shield = 1
            print('You equipped your shield!')
        else:
            char.shield = 0
            print('You unequipped your shield!')

    elif choice == '3':
        if data.bandage['amount'] > 0:
            print('You healed 10 hitpoints!')
            char.player.heal(10)
            data.bandage['amount'] -= 1
            bandage = data.bandage['amount']
            if bandage == 0:
                        data.inv.remove('bandage')
            print(f'You have {bandage} bandages remaining!')

        else:
            print('You don\'t have any more bandages')

        if char.player.health > char.player.maxHealth:
            char.player.health = char.player.maxHealth

    elif choice == '4' or choice.lower() == 'quit':
        exit()

    else:
        print('That is not a valid option!')
        fight()


def inCombat():

    global death

    # combat against a spider
    if char.combat == 1:
        print('\nYou are up against a spider')
        print(f'It has {classes.spider.health} hitpoints remaining\n')
        fight()
        if char.shield == 1:
            char.player.health -= (data.spider['damage']) / 2

        else:
            char.player.health -= (data.spider['damage'])

        if classes.spider.health <= 0:
            char.combat = 0
            classes.spider.kill()
            if char.player.health > 0:
                print('You squashed the spider!')
            if char.shield == 1:
                char.shield = 0
                print('Your shield has been unequipped')
            death += 1

# combat against an alien
    if char.combat == 2:
        print('\nYou are up against an alien')
        print(f'It has {classes.alien.health} hitpoints remaining\n')
        fight()
        if char.shield == 1:
            char.player.health -= (data.alien['damage']) / 2

        else:
            char.player.health -= (data.alien['damage'])

        if classes.alien.health <= 0:
            classes.alien.kill()
            if char.player.health > 0:
                print('The alien has been slain!')
            if char.shield == 1:
                char.shield = 0
                print('Your shield has been unequipped')
            char.combat = 0
            death += 1


# combat against a zombie
    if char.combat == 3:
        print('\nYou are up against a zombie')
        print(f'It has {classes.zombie.health} hitpoints remaining\n')
        fight()
        if char.shield == 1:
            char.player.health -= (data.zombie['damage']) / 2

        else:
            char.player.health -= (data.zombie['damage'])

        if classes.zombie.health <= 0:
            classes.zombie.kill()
            if char.player.health > 0:
                print('That zombie won\'t be needing any more brains!')
            char.combat = 0
            if char.shield == 1:
                char.shield = 0
                print('Your shield has been unequipped')
            death += 1
